### PR TITLE
[fix] [broker] Remove the repl cursor if the creation fails and the replication is turned off

### LINKED ISSUE #20010: Flaky-test: PersistentTopicTest.testCreateTopicWithZombieReplicatorCursor
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Example failure

https://github.com/apache/pulsar/actions/runs/4604351215/jobs/8140558256?pr=20005#step:11:1177

### Exception stacktrace

```
  Error:  Tests run: 82, Failures: 1, Errors: 0, Skipped: 77, Time elapsed: 48.113 s <<< FAILURE! - in org.apache.pulsar.broker.service.persistent.PersistentTopicTest
  Error:  testCreateTopicWithZombieReplicatorCursor(org.apache.pulsar.broker.service.persistent.PersistentTopicTest)  Time elapsed: 0.367 s  <<< FAILURE!
  java.util.concurrent.ExecutionException: java.lang.RuntimeException: org.apache.pulsar.metadata.api.MetadataStoreException$NotFoundException: remote
  	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
  	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
  	at org.apache.pulsar.broker.service.persistent.PersistentTopicTest.testCreateTopicWithZombieReplicatorCursor(PersistentTopicTest.java:587)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at java.base/java.lang.Thread.run(Thread.java:833)
  Caused by: java.lang.RuntimeException: org.apache.pulsar.metadata.api.MetadataStoreException$NotFoundException: remote
  	at org.apache.pulsar.broker.service.BrokerService.lambda$getReplicationClient$47(BrokerService.java:1347)
  	at org.apache.pulsar.common.util.collections.ConcurrentOpenHashMap$Section.put(ConcurrentOpenHashMap.java:409)
  	at org.apache.pulsar.common.util.collections.ConcurrentOpenHashMap.computeIfAbsent(ConcurrentOpenHashMap.java:243)
  	at org.apache.pulsar.broker.service.BrokerService.getReplicationClient(BrokerService.java:1271)
  	at org.apache.pulsar.broker.service.persistent.PersistentTopic.lambda$addReplicationCluster$65(PersistentTopic.java:1719)
  	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
  	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
  	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.handleGetResult(ZKMetadataStore.java:269)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.lambda$batchOperation$5(ZKMetadataStore.java:219)
  	at org.apache.zookeeper.MockZooKeeper.multi(MockZooKeeper.java:1006)
  	at org.apache.zookeeper.MockZooKeeperSession.multi(MockZooKeeperSession.java:191)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.batchOperation(ZKMetadataStore.java:190)
  	at org.apache.pulsar.metadata.impl.batching.AbstractBatchedMetadataStore.internalBatchOperation(AbstractBatchedMetadataStore.java:184)
  	at org.apache.pulsar.metadata.impl.batching.AbstractBatchedMetadataStore.flush(AbstractBatchedMetadataStore.java:103)
  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
  	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
  	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  	... 1 more
  Caused by: org.apache.pulsar.metadata.api.MetadataStoreException$NotFoundException: remote
  	at org.apache.pulsar.broker.service.BrokerService.lambda$getReplicationClient$46(BrokerService.java:1274)
  	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
  	at org.apache.pulsar.broker.service.BrokerService.lambda$getReplicationClient$47(BrokerService.java:1274)
  	... 21 more
```

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by poorbarcode**
Fixes #20010

### Motivation

The Cursor`repl` is automatically created when replication is turned on and it will be deleted when replication is turned off. But there has a bug: if the Topic is not online when you turn off replication, the cursor cannot be deleted, and #19972 fixes this bug.

But #19972 cannot solve the following scenario that cause the flaky test #20010:
1. a replication task starting
2. get clusters( responses `[local_cluster, remote_cluster]`)
3. create cursor `repl`
4. at this moment, the `remote_cluster` is removed, and the remote cluster is down now.
5. then the task `start replicator` will fail, but the cursor still there

### Modifications

Differing with #19972, do remove the `repl` cursor when the task `start replicator` is failed.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: 
- https://github.com/poorbarcode/pulsar/pull/84

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java [hunk 1]
```diff
@@ -558,10 +559,10 @@ public void testCreateTopicWithZombieReplicatorCursor(boolean topicLevelPolicy)
         admin.tenants().updateTenant("prop", tenantInfo);
 
         if (topicLevelPolicy) {
-            admin.topics().setReplicationClusters(topicName, Collections.singletonList(remoteCluster));
+            admin.topics().setReplicationClusters(topicName, Arrays.asList(remoteCluster, "test"));
```
Comment (BewareMyPower) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:562]: Why did you add another cluster "test" to this test? Even without the change of `PersistentTopic`, this test could still pass.

Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:562]: - If we disabled replication, the `clusters` will contain one cluster `test`. See: https://github.com/apache/pulsar/blob/master/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerTestBase.java#L55-L56
- If we enabled replication, the `clusters1 will contain two clusters: local cluster and remote cluster

If the cluster is not set correctly, it can mislead others: `clusters` is an empty collection when Replication is disabled.

Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:562]: I think it's another topic about the difference determined by whether the replication is enabled. It does not affect this test.

**Comment by BewareMyPower** [comment]
It's still flaky.

image

```
java.util.concurrent.ExecutionException: java.lang.RuntimeException: org.apache.pulsar.metadata.api.MetadataStoreException$NotFoundException: remote

	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
	at org.apache.pulsar.broker.service.persistent.PersistentTopicTest.testCreateTopicWithZombieReplicatorCursor(PersistentTopicTest.java:587)
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java (hunk 1 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:562]: > I think it's another topic about the difference determined by whether the replication is enabled. It does not affect this test.

Correct.

**Comment by BewareMyPower** [comment]
Could you also change the title? The PR title is duplicated with #19972.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java (hunk 1 above)
Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java]: So, it's better to revert this unrelated change.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java [hunk 2]
```diff
@@ -577,15 +586,21 @@ public void testCreateTopicWithZombieReplicatorCursor(boolean topicLevelPolicy)
...
-        topic.initialize().get(3, TimeUnit.SECONDS);
-        Awaitility.await().atMost(3, TimeUnit.SECONDS)
-                .until(() -> !topic.getManagedLedger().getCursors().iterator().hasNext());
+        // Verify:
+        // 1. Topic can load success( use "topic.initialize()" instead of "load topic" ).
+        //    If the topic loading by client is failed, it will retry, so we can do retry "topic.initialize()".
+        // 2. The repl cursor will be deleted.
+        Awaitility.await().atMost(10, TimeUnit.SECONDS)
+                .until(() -> {
+                    topic.initialize().get(3, TimeUnit.SECONDS);
```
Comment (BewareMyPower) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:602]: It's meaningless because `topic.initialize` might throw an exception and it won't be retried.

```

java.util.concurrent.ExecutionException: java.lang.RuntimeException: org.apache.pulsar.metadata.api.MetadataStoreException$NotFoundException: remote

	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
	at org.apache.pulsar.broker.service.persistent.PersistentTopicTest.lambda$testCreateTopicWithZombieReplicatorCursor$10(PersistentTopicTest.java:602)
	at org.awaitility.core.CallableCondition$ConditionEvaluationWrapper.eval(CallableCondition.java:99)
```

Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:602]: perhaps `.ignoreExceptionsInstanceOf(MetadataStoreException.NotFoundException.class)` could help?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 3]
```diff
@@ -1706,37 +1706,53 @@ private CompletableFuture<Boolean> checkReplicationCluster(String remoteCluster)
...
+            }).whenComplete((ignore, ex) -> {
+                if (ex == null){
+                    replicationStartFuture.complete(null);
+                } else {
+                    checkReplicationCluster(remoteCluster).thenCompose(clusterExists -> {
+                        if (!clusterExists) {
+                            log.warn("[{}] Start remove the replicator because the cluster '{}' does not exist",
+                                    topic, remoteCluster);
+                            replicationStartFuture.complete(null);
+                            return removeReplicator(remoteCluster).whenComplete((ignore2, remoteCursorEx) -> {
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1738]: why is the replicationStartFuture completed before removing the replicator?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java [hunk 4]
```diff
@@ -558,10 +559,18 @@ public void testCreateTopicWithZombieReplicatorCursor(boolean topicLevelPolicy)
...
         if (topicLevelPolicy) {
-            admin.topics().setReplicationClusters(topicName, Collections.singletonList(remoteCluster));
+            // setReplicationClusters may fail, so do retry.
+            Awaitility.await().until(() -> {
+                try {
+                    admin.topics().setReplicationClusters(topicName, Arrays.asList(remoteCluster, "test"));
+                    return true;
+                } catch (Exception ex) {
+                    return false;
+                }
```
Comment (lhotari) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:569]: You could use `Awaitility.await().ignoreExceptions().untilAsserted(` instead

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java (hunk 2 above)
Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:602]: BTW, the root cause is not here. I added a `try-catch` block to swallow the exception, the test could still fail.

```java
            try {
                topic.initialize().get(3, TimeUnit.SECONDS);
            } catch (ExecutionException e) {
                log.warn("Failed to initialize: {}", e.getCause().getMessage());
            }
```

Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:602]: > BTW, the root cause is not here. I added a try-catch block to swallow the exception, the test could still fail.

The root cause is like this:
1. a replication task starting
2. failed to get cluster meta from ZK
3. check replication policies
  - Since the policy is updated asynchronously, the cache has not been updated.
  - get clusters( responses `[local_cluster, remote_cluster]`)
4. Since check replication successes, will not delete the cursor. bingo

Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:602]: I've run this test 4*200 times with no errors

截屏2023-04-06 21 04 25

Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:602]: >  just mock the load operation by client,

Actually, the operation failed at `topic.initialize` in my local env. An exception will be thrown and it won't retry.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 3 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Already fixed. Thanks

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java (hunk 4 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java]: Good idea. Thanks, already change to `Awaitility.await().ignoreExceptions().untilAsserted`

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java (hunk 2 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java]: See step 4 above, it goes to the logic branch `else`, and the error is thrown.

https://github.com/apache/pulsar/pull/20025/files#diff-5edf14cc6f25857d0cfdd26b2d3b3141230ecfb0dfa95aebf7583fd76ede4c4bR1748-R1751

Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java]: I have change the test to `Awaitility.await().atMost(10, TimeUnit.SECONDS).ignoreExceptions().untilAsserted(`, it will be fixed

**Comment by BewareMyPower** [comment]
@poorbarcode Could you also change the title? The PR title is duplicated with https://github.com/apache/pulsar/pull/19972.

**Comment by poorbarcode** [comment]
> Could you also change the title? The PR title is duplicated with https://github.com/apache/pulsar/pull/19972.

Done

**Comment by BewareMyPower** [comment]
Here is a failure from my local env: [failure.log](https://github.com/apache/pulsar/files/11171288/failure.log)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 5]
```diff
@@ -1706,37 +1706,54 @@ private CompletableFuture<Boolean> checkReplicationCluster(String remoteCluster)
...
+                    return null;
+                });
+                // clean up replicator if startup is failed
+                if (replicator == null) {
+                    replicators.removeNullValue(remoteCluster);
+                }
+            }).whenComplete((ignore, ex) -> {
+                if (ex == null){
+                    replicationStartFuture.complete(null);
+                } else {
```
Comment (BewareMyPower) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1732]: It's better to print the `ex` warning here.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java (hunk 1 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java]: reverted thse changes

**Comment by BewareMyPower** [comment]
After discussing with @poorbarcode offline, let's move to https://github.com/apache/pulsar/pull/20037 to fix the flaky test.

**Comment by michaeljmarshall** [comment]
As discussed on the mailing list https://lists.apache.org/thread/w4jzk27qhtosgsz7l9bmhf1t7o9mxjhp, there is no plan to release 2.9.6, so I am going to remove the release/2.9.6 label
